src/ir_passes/lower/compile.rs

```diff
@@ -0,0 +1,300 @@
+            attributes,
+        }
+    }
+
+    /// Compiles a list of ports into a [calyx::PortDef].
+    fn ports<CW, WFU, WT>(
+        ctx: &ir::Context,
+        comp: &ir::Component,
+        width_from_u64: WFU, // Function that returns a CW type from a u64
+        width_transform: WT, // Function that transforms an [ir::ExprIdx] into a [CW] type
+    ) -> Vec<calyx::PortDef<CW>>
+    where
+        WFU: Fn(u64) -> CW,
+        WT: Fn(&ir::Component, ir::ExprIdx) -> CW,
+    {
+        let mut ports: Vec<_> = comp
+            // the initial list of ports.
+            .ports()
+            .idx_iter()
+            .filter(|idx| comp.get(*idx).is_sig())
+            .map(|idx| Compile::port_def(ctx, comp, idx, &width_transform))
+            .chain(
+                // adds unannotated ports to the list of ports
+                comp.unannotated_ports().into_iter().map(|(name, width)| {
+                    calyx::PortDef {
+                        name: name.as_ref().into(),
+                        width: width_from_u64(width),
+                        direction: calyx::Direction::Input,
+                        attributes: calyx::Attributes::default(),
+                    }
+                }),
+            )
+            .chain(
+                // adds interface ports to the list of ports
+                comp.events()
+                    .idx_iter()
+                    .filter_map(|idx| idx.interface_name(comp))
+                    .into_iter()
+                    .map(|name| calyx::PortDef {
+                        name: name.into(),
+                        width: width_from_u64(1),
+                        direction: calyx::Direction::Input,
+                        // adds the `@fil_event` attribute to the port
+                        attributes: vec![(
+                            calyx::Attribute::Unknown("fil_event".into()),
+                            1,
+                        )]
+                        .try_into()
+                        .unwrap(),
+                    }),
+            )
+            .collect();
+
+        let mut interface_ports =
+            INTERFACE_PORTS.iter().collect::<HashSet<_>>();
+
+        // add interface port attributes if necessary
+        for pd in &mut ports {
+            if let Some(pair @ ((attr, value), (name, _, dir))) =
+                INTERFACE_PORTS
+                    .iter()
+                    .find(|(_, (n, _, _))| *n == pd.name.as_ref())
+            {
+                assert!(
+                    dir == &pd.direction,
+                    "Expected {} to be an {:?} port, got {:?} port.",
+                    name,
+                    dir,
+                    pd.direction
+                );
+                // TODO: should also assert that the width of the matching port is the same as what we expect
+                // We'd also need an equality function on `CW` types to do this, which we don't have at the moment
+                // for [calyx::Width].
```

Comment:
Oh, maybe we should open an issue for the Calyx repo?

Reply:
Will do